### PR TITLE
Modifications to allow Serialbox to be used in CMake superbuild projects

### PR DESCRIPTION
**Pull request by wdeconinck**
This PR allows Serialbox to be used in CMake superbuild projects,
so that one could have code like
```
add_subdirectory( serialbox2 )  # --> will set "Serialbox_DIR" in cache
add_subdirectory( project_using_serialbox )
```
and within `project_using_serialbox` we have
```
find_package( Serialbox )  # --> will use Serialbox_DIR as default hint
```
Most issues were related to use of `${CMAKE_BINARY_DIR}` and `${CMAKE_SOURCE_DIR}` which reflect the top-level CMake project rather than `${PROJECT_BINARY_DIR}` and `${PROJECT_SOURCE_DIR}` which reflect the most recent project.

As a result it could be that certain products are no longer in a previously expected directory now, as I am not able to test everything

---
Moreover in this PR I have created alias targets (if possible) `Serialbox::Serialbox_C` and `Serialbox::Serialbox_Fortran` that prefer the shared versions `Serialbox::SerialboxCShared` and `Serialbox::SerialboxFortranShared` over the static versions `Serialbox::SerialboxCStatic` and `Serialbox::SerialboxFortranStatic`

In the client software you can then just use the alias target instead.

### CODE REVIEW

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

File: cmake/SerialboxConfig.cmake.in [hunk 1]
```diff
@@ -40,12 +40,78 @@ set(SERIALBOX_BOOST_LIBRARY_DIRS "@Boost_LIBRARY_DIRS@")
...
+  endif()
+endfunction()
+
 #===---------------------------------------------------------------------------------------------===
 #   Find Serialbox libraries
 #====--------------------------------------------------------------------------------------------===
 
 # Import library targets 
-include("${CMAKE_CURRENT_LIST_DIR}/SerialboxTargets.cmake")
+if( NOT @PROJECT_NAME@_TARGETS_EXPORTED )
```
Comment (havogt) [inline cmake/SerialboxConfig.cmake.in]: Why is this protection necessary?

File: cmake/SerialboxConfig.cmake.in [hunk 2]
```diff
@@ -40,12 +40,78 @@ set(SERIALBOX_BOOST_LIBRARY_DIRS "@Boost_LIBRARY_DIRS@")
...
+  #
+  #    Creates ALIAS target for given targets. Prefer target_shared if availabe
+  # 
+  set( _alias ${ARGV0} )
+  set( _target_shared ${ARGV1} )
+  set( _target_static ${ARGV2} )
+  set( _align ${ARGV3} )
+  string( REPLACE "Serialbox::" "" _target_shared_build "${_target_shared}" )
+  string( REPLACE "Serialbox::" "" _target_static_build "${_target_static}" )
+  if( TARGET ${_target_shared_build} )
```
Comment (havogt) [inline cmake/SerialboxConfig.cmake.in]: Why is this necessary? You also added the `serialbox_install_targets()` which should have been called in the case of `if( TARGET ${_target_shared_build} )`, so no need to check here if the alias exists?

File: cmake/SerialboxConfig.cmake.in (hunk 1 above)
Comment (wdeconinck) [reply cmake/SerialboxConfig.cmake.in]: If you are in a nested CMake tree like:
```
super-cmake-project/
    serialbox2/
    myproject/
```
and `myproject` has a `find_package(SerialBox)` then the targets are already globally visible, so we must avoid including the targets again. the `@PROJECT_NAME@_TARGETS_EXPORTED` variable is set during the configuration of `serialbox2`

File: cmake/SerialboxConfig.cmake.in (hunk 2 above)
Comment (wdeconinck) [reply cmake/SerialboxConfig.cmake.in]: I wanted to create the aliases
- Serialbox::Serialbox_C   --> SerialboxCShared or SerialboxCStatic
- Serialbox::Serialbox_Fortran --> SerialboxFortranShared or SerialboxFortranStatic

which prefer the shared build if available, and fall-back to static. It's always better to use shared because Serialbox2 dependencies can stay private. Using the aliases, client code needs not to analyze if shared is available. And if static is what you wanted you can always link to Serialbox::SerialboxCStatic as it is always available

For `Serialbox::Serialbox_C`:

If 1) `SerialboxCShared` or 2) `Serialbox::SerialboxCShared` are a target then an alias is created if not available yet.

In case of 1) we are in a super-build configuration and the target `Serialbox::SerialboxCShared` target exists. But CMake disallows to alias an alias. Therefore we need to alias the original target
```
Serialbox::SerialboxC --> SerialboxCShared
```
In case of 2) Serialbox is from a separate install-tree.  `SerialboxCShared` does not exist as target, but `Serialbox::SerialboxCShared` exists as **IMPORTED** target. It ***is*** allowed to alias an IMPORTED target, provided that we set the PROPERTY ```IMPORTED_GLOBAL```
